### PR TITLE
Add IndependentFactoredGenerativeProcess for frozen factor support

### PR DESCRIPTION
**Pull request by loren-ac**
### Summary

Introduces `IndependentFactoredGenerativeProcess`, a subclass of `FactoredGenerativeProcess` that:
- Samples emissions from each factor **independently** (rather than from the joint distribution)
- Supports **frozen factors** whose emission sequences are identical across all batch samples

This enables generating datasets where k of n factors share the same realizations across samples while the remaining (n-k) factors vary independently.

### Motivation

When working with factored generative processes, it's useful to control which factors are held constant vs. which vary across a batch of generated sequences. This is particularly valuable for:
- Ablation studies isolating the effect of specific factors
- Generating controlled datasets with shared structure across samples
- Investigating factor-specific properties of sequence prediction models

### Implementation

The key mechanism is a **dual key stream** approach in `generate()`:
- **Frozen factors**: Use keys derived from a stored `frozen_key` (same across all batch samples → identical sequences)
- **Unfrozen factors**: Use keys derived from the per-sample key (different per sample → varying sequences)

```python
process = IndependentFactoredGenerativeProcess(
    component_types=("hmm", "hmm", "hmm"),
    transition_matrices=...,
    normalizing_eigenvectors=...,
    initial_states=...,
    structure=IndependentStructure(),
    frozen_factor_indices=frozenset({0, 2}),  # Factors 0 and 2 are frozen
    frozen_key=jax.random.PRNGKey(42),
)

# Factors 0 and 2 will be identical across all batch samples
states, observations = process.generate(batch_states, keys, seq_len, False)
```

### Changes

- New file: simplexity/generative_processes/independent_factored_generative_process.py
- New tests: tests/generative_processes/test_independent_factored_generative_process.py (13 tests, 100% coverage)

### Test plan

- All existing FactoredGenerativeProcess tests pass (no regressions)
- New tests verify frozen factor consistency across batch
- New tests verify unfrozen factor variation across batch
- New tests verify reproducibility of frozen sequences across generate() calls
- Edge cases covered: all frozen, none frozen, invalid indices
- Type checking passes (pyright)
- Linting passes (ruff)

### CODE REVIEW

File: simplexity/generative_processes/independent_factored_generative_process.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+                factor_key = frozen_factor_keys[i]
+            else:
+                factor_key = factor_keys[i]
+
+            T_i = self.transition_matrices[i][0]
+            norm_i = self.normalizing_eigenvectors[i][0] if self.component_types[i] == "ghmm" else None
+            p_i = compute_obs_dist_for_variant(self.component_types[i], state[i], T_i, norm_i)
+
+            token_i = jax.random.categorical(factor_key, jnp.log(p_i))
+            per_factor_tokens.append(token_i)
```
[the remark this answers is not in the document]
Comment (loren-ac) [reply simplexity/generative_processes/independent_factored_generative_process.py:129]: Claude code claims that this feedback is wrong:

```
That Copilot feedback is actually incorrect for this specific case. Here's why the current code is fine:

  1. The Python loop is unrolled at trace time

  Since num_factors = len(self.component_types) is a static Python integer, JAX unrolls the for i in range(num_factors) loop during tracing. The loop doesn't exist at runtime - it becomes a flat sequence of operations.

  2. The conditional is resolved at trace time

  if i in self.frozen_factor_indices involves:
  - i - a Python integer from range()
  - self.frozen_factor_indices - a Python frozenset

  Both are Python values, not JAX arrays, so the conditional is evaluated during tracing, not execution. JAX traces different code paths depending on which factors are frozen.

  3. List append is fine during tracing

  The list is built during tracing, then converted to a tuple. The final traced function has no list operations - just the unrolled operations.

  ---
  Using jax.lax.fori_loop would actually be worse here because:
  - It would require stacking heterogeneous factor data (different state dimensions per factor)
  - It would force runtime conditionals instead of compile-time specialization
  - The loop would execute at runtime instead of being unrolled
  ```

File: simplexity/generative_processes/independent_factored_generative_process.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            raise ValueError("frozen_key is required when frozen_factor_indices is non-empty")
+
+        if not isinstance(structure, IndependentStructure):
+            SIMPLEXITY_LOGGER.warning(
+                "IndependentFactoredGenerativeProcess is designed for IndependentStructure. "
+                "Using %s may produce unexpected results.",
+                type(structure).__name__,
+            )
+
+        self.frozen_factor_indices = frozen_factor_indices
```
[the remark this answers is not in the document]
Comment (loren-ac) [reply simplexity/generative_processes/independent_factored_generative_process.py:98]: Working on a fix for this now

**Review by hrbigelow — APPROVED** [review]
Main thing I wonder is, how do you update the frozen key so that you can get multiple settings for the frozen trajectories.  I guess you can do this in the main loop?

```python
for b in range(num_batches):
    batch = ds.generate()
    if b % 10 == 0:
      ds.frozen_key = jax.random.fold_in(ds.frozen_key, b)
```

something like that?

**Review by ealt — COMMENTED** [review]
What do you think of this refactor?

https://github.com/Astera-org/simplexity/pull/169
